lib/matplotlib/axes/_axes.py

```diff
@@ -2381,6 +2381,14 @@ def bar(self, x, height, width=0.8, bottom=None, *, align="center",
             tick_label_axis = self.yaxis
             tick_label_position = y
 
+        patch_labels = np.atleast_1d(label)
+        if len(patch_labels) != len(x):
+            if len(patch_labels) == 1:
+                patch_labels = ['_nolegend_'] * len(x)
+            else:
+                raise ValueError(f'number of labels ({len(patch_labels)}) '
+                                 f'does not match number of bars ({len(x)}).')
```

Comment:
Very subtle, but I think this is better:
```suggestion
        if np.iterable(labels) and not isinstance(labels, str):
            patch_labels = labels
        else:
            ['_nolegend_'] * len(x)
        if len(patch_labels) != len(x):
            raise ValueError(f'number of labels ({len(patch_labels)}) '
                             f'does not match number of bars ({len(x)}).')

```

Using `np.atleast_1d(labels)` makes `label=['Apple']` behave like `label='Apple'` (and that would be accepted.
The suggested version raises on `label=['Apple']` if there are multiple x values.

Please add a test case for this case.